File: tests/test_command_edit.py
# pyright: standard

import os
import subprocess
from pathlib import Path
from unittest.mock import MagicMock

import pytest
from pytest_mock import MockerFixture
from typer.testing import CliRunner

from aico.lib.models import (
    AssistantChatMessage,
    DerivedContent,
    Mode,
    SessionData,
    UserChatMessage,
)
from aico.lib.session import SESSION_FILE_NAME, save_session
from aico.main import app
from tests.test_command_undo import load_session_data

runner = CliRunner()


@pytest.fixture
def session_with_context_file(tmp_path: Path) -> Path:
    """Creates a session file in a temporary directory with one file in context."""
    session_dir = tmp_path
    os.chdir(session_dir)
    (session_dir / "code.py").write_text("def hello(): pass")
    session_file = session_dir / SESSION_FILE_NAME
    session_data = SessionData(
        model="test-model",
        context_files=["code.py"],
        chat_history=[],
    )
    save_session(session_file, session_data)
    return session_file


def mock_editor(new_content: str, return_code: int = 0) -> MagicMock:
    """Creates a mock side effect for subprocess.run to simulate an editor."""

    def _side_effect(cmd_parts: list[str], check: bool) -> subprocess.CompletedProcess[str]:  # pyright: ignore[reportUnusedParameter]
        temp_file_path = Path(cmd_parts[-1])
        if return_code == 0:
            temp_file_path.write_text(new_content)
        return subprocess.CompletedProcess(args=cmd_parts, returncode=return_code, stdout="", stderr="")

    return MagicMock(side_effect=_side_effect)


def test_edit_prompt(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a session, a mocked editor, and an environment where EDITOR is not set
    mocker.patch.dict(os.environ).pop("EDITOR", None)

    session_file = session_with_two_pairs
    mock_run = mocker.patch("subprocess.run", new=mock_editor("Updated prompt content"))

    # WHEN `aico edit --prompt` is run on the first pair (index 0)
    result = runner.invoke(app, ["edit", "0", "--prompt"])

    # THEN the command succeeds and reports the update
    assert result.exit_code == 0, result.stderr
    assert "Updated prompt for message pair 0." in result.stdout

    # AND the session file is updated with the new prompt content
    final_session = load_session_data(session_file)
    assert final_session.chat_history[0].content == "Updated prompt content"
    assert final_session.chat_history[1].content == "assistant response 0"  # Unchanged

    # AND the editor was called correctly
    mock_run.assert_called_once()
    assert mock_run.call_args[0][0][0] == "vi"  # Default editor


def test_edit_response_and_invalidate_derived_content(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a session where the last response has derived content
    session_file = session_with_two_pairs
    session_data = load_session_data(session_file)
    from dataclasses import replace

    last_response = session_data.chat_history[-1]
    last_response_with_derived = replace(
        last_response, derived=DerivedContent(unified_diff="diff", display_content="display")
    )
    session_data.chat_history[-1] = last_response_with_derived
    save_session(session_file, session_data)

    # AND a mocked editor that will succeed
    mock_run = mocker.patch("subprocess.run", new=mock_editor("Updated response content"))

    # WHEN `aico edit` is run with default arguments (last response)
    result = runner.invoke(app, ["edit"])

    # THEN the command succeeds
    assert result.exit_code == 0
    assert "Updated response for message pair 1." in result.stdout

    # AND the response content is updated
    final_session = load_session_data(session_file)
    final_response = final_session.chat_history[-1]
    assert final_response.content == "Updated response content"

    # AND the derived content is recomputed but is None because the new content is just conversational
    assert isinstance(final_response, AssistantChatMessage)
    assert final_response.derived is None

    mock_run.assert_called_once()


def test_edit_with_custom_editor(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a mocked editor and a custom EDITOR env var
    _ = session_with_two_pairs
    mock_run = mocker.patch("subprocess.run", new=mock_editor("new content"))
    mocker.patch.dict(os.environ, {"EDITOR": "nvim -f"})

    # WHEN `aico edit` is run
    result = runner.invoke(app, ["edit"])

    # THEN the command succeeds
    assert result.exit_code == 0

    # AND the custom editor command was used
    mock_run.assert_called_once()
    called_cmd = mock_run.call_args[0][0]
    assert called_cmd[0] == "nvim"
    assert called_cmd[1] == "-f"


def test_edit_aborts_if_no_changes(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a mocked editor that makes no changes
    session_file = session_with_two_pairs
    original_content = load_session_data(session_file).chat_history[-1].content
    mocker.patch("subprocess.run", new=mock_editor(original_content))

    # WHEN `aico edit` is run
    result = runner.invoke(app, ["edit"])

    # THEN the command succeeds but reports no changes
    assert result.exit_code == 0
    assert "No changes detected. Aborting." in result.stdout


def test_edit_aborts_on_editor_failure(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a mocked editor that returns a non-zero exit code
    _ = session_with_two_pairs
    mocker.patch("subprocess.run", new=mock_editor("irrelevant", return_code=1))

    # WHEN `aico edit` is run
    result = runner.invoke(app, ["edit"])

    # THEN the command fails and reports the editor failure
    assert result.exit_code == 1
    assert "Editor closed with non-zero exit code. Aborting." in result.stderr


def test_edit_fails_on_bad_index(session_with_two_pairs: Path) -> None:  # noqa F811 module-private
    # GIVEN a session
    # WHEN `aico edit` is called with an out-of-bounds index
    result = runner.invoke(app, ["edit", "99"])

    # THEN it fails with a clear error
    assert result.exit_code == 1
    assert "Error: Pair at index 99 not found." in result.stderr


def test_edit_fails_if_editor_not_found(session_with_two_pairs: Path, mocker: MockerFixture) -> None:  # noqa F811 module-private
    # GIVEN a mocked subprocess.run that raises FileNotFoundError
    mocker.patch("subprocess.run", side_effect=FileNotFoundError)

    # WHEN `aico edit` is run
    result = runner.invoke(app, ["edit"])

    # THEN the command fails with a helpful error message
    assert result.exit_code == 1
    assert "Error: Editor command not found" in result.stderr
    assert "Please set the $EDITOR environment variable." in result.stderr


def test_edit_response_recomputes_derived_content_on_change(
    session_with_context_file: Path, mocker: MockerFixture
) -> None:
    # GIVEN a session with a file in context and one message pair
    session_file = session_with_context_file
    session_data = load_session_data(session_file)

    user_msg = UserChatMessage(role="user", content="prompt", mode=Mode.CONVERSATION, timestamp="ts")
    assistant_msg = AssistantChatMessage(
        role="assistant",
        content="Initial conversational response.",
        mode=Mode.CONVERSATION,
        timestamp="ts",
        model="test-model",
        duration_ms=100,
    )
    session_data.chat_history = [user_msg, assistant_msg]
    save_session(session_file, session_data)

    # AND a mocked editor that will provide a valid patch
    new_patch_content = (
        "File: code.py\n<<<<<<< SEARCH\ndef hello(): pass\n=======\ndef greeting(): print('hello')\n>>>>>>> REPLACE"
    )
    mock_run = mocker.patch("subprocess.run", new=mock_editor(new_patch_content))

    # WHEN `aico edit` is run with default arguments (last response)
    result = runner.invoke(app, ["edit"])

    # THEN the command succeeds
    assert result.exit_code == 0, result.stderr
    assert "Updated response for message pair 0." in result.stdout

    # AND the response content is updated
    final_session = load_session_data(session_file)
    final_response = final_session.chat_history[-1]
    assert isinstance(final_response, AssistantChatMessage)
    assert final_response.content == new_patch_content

    # AND the derived content is recomputed and contains the new diff
    assert final_response.derived is not None
    assert isinstance(final_response.derived, DerivedContent)

    expected_diff = (
        "--- a/code.py\n"
        "+++ b/code.py\n"
        "@@ -1 +1 @@\n"
        "-def hello(): pass\n"
        "\\ No newline at end of file\n"
        "+def greeting(): print('hello')\n"
    )
    assert final_response.derived.unified_diff == expected_diff

    # AND the display content is also correct
    expected_display_items = [
        {"type": "markdown", "content": "File: `code.py`\n"},
        {"type": "diff", "content": expected_diff},
    ]
    assert final_response.derived.display_content == expected_display_items
    mock_run.assert_called_once()

File: tests/test_command_edit.rs
mod common;
use assert_cmd::cargo::cargo_bin_cmd;
use common::init_session_with_history;
use predicates::prelude::*;
use std::env;
use std::fs;
use std::os::unix::fs::PermissionsExt;
use tempfile::tempdir;

#[test]
fn test_edit_prompt_success() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("original prompt", "r0")]);

    // Use a shell script as faker editor that replaces content
    let editor_script = root.join("editor.sh");
    fs::write(&editor_script, "#!/bin/sh\necho 'edited prompt' > \"$1\"").unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = fs::metadata(&editor_script).unwrap().permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&editor_script, perms).unwrap();
    }

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .args(["edit", "0", "--prompt"])
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Updated prompt for message pair 0.",
        ));

    // Verify change in history via log
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("log")
        .assert()
        .success()
        .stdout(predicate::str::contains("edited prompt"));
}

#[test]
fn test_edit_response_recomputes_derived_content_on_change() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // GIVEN a session with a file and a conversational response
    fs::write(root.join("file.py"), "print('hello')\n").unwrap();
    common::init_session_with_history(root, vec![("Change it", "Sure.")]);
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["add", "file.py"])
        .assert()
        .success();

    // AND an editor that replaces "Sure." with a valid SEARCH/REPLACE block
    // The SEARCH block must match the file content exactly, including the newline.
    let new_content =
        "File: file.py\n<<<<<<< SEARCH\nprint('hello')\n=======\nprint('world')\n>>>>>>> REPLACE";
    let editor_script = root.join("editor.sh");
    fs::write(
        &editor_script,
        format!("#!/bin/sh\necho \"{}\" > \"$1\"", new_content),
    )
    .unwrap();
    fs::set_permissions(&editor_script, fs::Permissions::from_mode(0o755)).unwrap();

    // WHEN editing the response
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .success();

    // THEN the derived content (diff) should be recomputed and visible in 'last --json'
    let output = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--json"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();

    let json: serde_json::Value = serde_json::from_slice(&output).unwrap();
    let asst = &json["assistant"];
    let diff = asst["derived"]["unified_diff"]
        .as_str()
        .expect("unified_diff should be present in derived content after edit");
    assert!(diff.contains("-print('hello')"));
    assert!(diff.contains("+print('world')"));

    // AND display items should be structured
    let display_items = asst["derived"]["display_content"]
        .as_array()
        .expect("display_content should be an array");
    assert!(display_items.iter().any(|item| {
        item["type"] == "diff"
            && item["content"]
                .as_str()
                .unwrap()
                .contains("-print('hello')")
    }));
}

#[test]
fn test_edit_with_custom_editor() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    common::init_session_with_history(root, vec![("p", "r")]);

    // Use a custom string for EDITOR with flags
    let editor_script = root.join("my_editor.sh");
    fs::write(&editor_script, "#!/bin/sh\necho \"custom editor\" > \"$2\"").unwrap();
    fs::set_permissions(&editor_script, fs::Permissions::from_mode(0o755)).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", format!("{} -f", editor_script.to_str().unwrap()))
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("custom editor"));
}

#[test]
fn test_edit_aborts_if_no_changes() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    // A faker editor that does nothing
    let editor_script = root.join("noop_editor.sh");
    fs::write(&editor_script, "#!/bin/sh\nexit 0").unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = fs::metadata(&editor_script).unwrap().permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&editor_script, perms).unwrap();
    }

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .arg("edit")
        .assert()
        .success()
        .stdout(predicate::str::contains("No changes detected. Aborting."));
}

#[test]
fn test_edit_fails_on_editor_error() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", "false")
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .failure()
        .stderr(predicate::str::contains("Editor closed with exit code"));
}

#[test]
fn test_edit_aborts_on_editor_failure() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    // editor-script that exits with 1
    let editor_script = root.join("fail_editor.sh");
    fs::write(&editor_script, "#!/bin/sh\nexit 1").unwrap();
    fs::set_permissions(&editor_script, fs::Permissions::from_mode(0o755)).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "Editor closed with exit code 1. Aborting.",
        ));
}

#[test]
fn test_edit_prompt() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    let editor_script = root.join("editor.sh");
    fs::write(
        &editor_script,
        "#!/bin/sh\necho 'new prompt content' > \"$1\"",
    )
    .unwrap();
    fs::set_permissions(&editor_script, fs::Permissions::from_mode(0o755)).unwrap();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .args(["edit", "--prompt"])
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--prompt"])
        .assert()
        .success()
        .stdout(predicate::str::contains("new prompt content"));
}

#[test]
fn test_edit_fails_if_editor_not_found() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", "/non/existent/editor/binary")
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .failure()
        .stderr(predicate::str::contains("not found. Please set $EDITOR."));
}

#[test]
fn test_edit_updates_store_and_view() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    // Use a shell script as faker editor
    let editor_script = root.join("editor.sh");
    fs::write(&editor_script, "#!/bin/sh\necho 'edited r0' > \"$1\"").unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = fs::metadata(&editor_script).unwrap().permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&editor_script, perms).unwrap();
    }

    // WHEN editing the assistant response (index 0, which corresponds to pair 0)
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .success();

    // THEN the view is updated to point to a new record ID
    let view = common::load_view(root);
    // Original indices were [0, 1]. New should be [0, 2]
    assert_eq!(view.message_indices.len(), 2);
    assert_eq!(view.message_indices[0], 0);
    assert!(view.message_indices[1] > 1);

    // AND the store contains both the original and the new edit
    let history_path = root.join(".aico/history/0.jsonl");
    let content = fs::read_to_string(history_path).unwrap();
    assert!(content.contains("r0"));
    assert!(content.contains("edited r0"));
    assert!(content.contains("\"edit_of\":1"));
}

#[test]
fn test_edit_response_and_invalidate_derived_content() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    // Manually inject derived content into the history record by parsing and re-serializing
    let history_path = root.join(".aico/history/0.jsonl");
    let content = fs::read_to_string(&history_path).unwrap();
    let lines: Vec<&str> = content.lines().collect();
    let mut rec: aico::models::HistoryRecord = serde_json::from_str(lines[1]).unwrap();

    rec.derived = Some(aico::models::DerivedContent {
        unified_diff: Some("some diff".to_string()),
        display_content: None,
    });

    let updated_line = serde_json::to_string(&rec).unwrap();
    fs::write(&history_path, format!("{}\n{}\n", lines[0], updated_line)).unwrap();

    // Verify derived content is there
    let output_before = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--json"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();
    assert!(
        String::from_utf8(output_before)
            .unwrap()
            .contains("some diff")
    );

    let editor_script = root.join("editor.sh");
    fs::write(
        &editor_script,
        "#!/bin/sh\necho 'conversational edit' > \"$1\"",
    )
    .unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = fs::metadata(&editor_script).unwrap().permissions();
        perms.set_mode(0o755);
        fs::set_permissions(&editor_script, perms).unwrap();
    }

    // WHEN editing (provide empty stdin to prevent hang in non-TTY test environment)
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_script.to_str().unwrap())
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .write_stdin("")
        .assert()
        .success();

    // THEN the new record in JSON output should NOT have derived content
    // because it was a simple text edit
    let output_after = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--json"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();
    let json_after: serde_json::Value = serde_json::from_slice(&output_after).unwrap();
    assert!(json_after["assistant"]["derived"].is_null());
}

#[test]
fn test_edit_scripted_mode() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "original response")]);

    // WHEN running aico edit in a non-tty environment (stdin is redirected)
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("edit")
        .write_stdin("new response from stdin")
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Updated response for message pair 0.",
        ));

    // THEN the response content is updated
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("new response from stdin"));
}

#[test]
fn test_edit_scripted_mode_empty_stdin() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "original response")]);

    // WHEN running aico edit with empty piped input
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("edit")
        .write_stdin("")
        .assert()
        .success()
        .stdout(predicate::str::contains("No changes detected. Aborting."));

    // THEN the response remains unchanged
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("original response"));
}

#[test]
fn test_edit_fails_on_bad_index() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0")]);

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["edit", "99"])
        .assert()
        .failure()
        .stderr(predicate::str::contains("Index out of bounds"));
}

#[test]
fn test_edit_negative_index() {
    let temp = tempfile::tempdir().unwrap();
    let root = temp.path();

    init_session_with_history(root, vec![("p0", "r0"), ("p1", "r1")]);

    // -1 highlights the last pair
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["edit", "-1"])
        .write_stdin("updated last")
        .assert()
        .success()
        .stdout(predicate::str::contains(
            "Updated response for message pair 1.",
        ));

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("updated last"));
}

#[test]
fn test_edit_preserves_original_timestamp() {
    let temp = tempdir().unwrap();
    let root = temp.path();

    // GIVEN a session with one pair
    common::init_session_with_history(root, vec![("p0", "r0")]);

    // AND we capture the original timestamp
    let output_before = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--json"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();
    let json_before: serde_json::Value = serde_json::from_slice(&output_before).unwrap();
    let original_ts = json_before["assistant"]["timestamp"].as_str().unwrap();

    // WHEN editing the response via piped input
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("edit")
        .write_stdin("edited response")
        .assert()
        .success();

    // THEN the timestamp remains unchanged
    let output_after = cargo_bin_cmd!("aico")
        .current_dir(root)
        .args(["last", "--json"])
        .assert()
        .success()
        .get_output()
        .stdout
        .clone();
    let json_after: serde_json::Value = serde_json::from_slice(&output_after).unwrap();
    let new_ts = json_after["assistant"]["timestamp"].as_str().unwrap();

    assert_eq!(original_ts, new_ts, "Timestamp was changed during edit!");
}

#[test]
fn test_edit_handles_editor_path_with_spaces() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    common::init_session_with_history(root, vec![("p", "original")]);

    // Create a directory with spaces and a script inside
    let editor_dir = root.join("my editor");
    fs::create_dir_all(&editor_dir).unwrap();
    let editor_script = editor_dir.join("edit.sh");

    // Script that writes to the last argument (the file path)
    fs::write(
        &editor_script,
        "#!/bin/sh\nfor last; do :; done\necho \"space editor\" > \"$last\"",
    )
    .unwrap();
    fs::set_permissions(&editor_script, fs::Permissions::from_mode(0o755)).unwrap();

    // Wrap the path in quotes to simulate a robust $EDITOR configuration
    let editor_val = format!("\"{}\" -f", editor_script.to_str().unwrap());

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", editor_val)
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("space editor"));
}

#[test]
fn test_edit_handles_editor_with_flags_string() {
    let temp = tempdir().unwrap();
    let root = temp.path();
    common::init_session_with_history(root, vec![("p", "original")]);

    // Set EDITOR to something that mimics a command with flags
    // Under Unix, we can use 'sed' to replace the content
    cargo_bin_cmd!("aico")
        .current_dir(root)
        .env("EDITOR", "sed -i s/original/modified/g")
        .env("AICO_FORCE_EDITOR", "1")
        .arg("edit")
        .assert()
        .success();

    cargo_bin_cmd!("aico")
        .current_dir(root)
        .arg("last")
        .assert()
        .success()
        .stdout(predicate::str::contains("modified"));
}
